lib/ansible/modules/packaging/os/snap.py

```diff
@@ -0,0 +1,204 @@
+              like those managed by APT. This option corresponds to the --classic argument.
+        type: bool
+        required: false
+        default: False
+
+author:
+    - Victor Carceler (vcarceler@iespuigcastellar.xeill.net)
+    - Stanislas Lange (angristan) <angristan@pm.me>
+'''
+
+EXAMPLES = '''
+# Install "foo" snap
+- name: Install foo
+  snap:
+    name: foo
+
+# Remove "foo" snap
+- name: Remove VLC
+  snap:
+    name: foo
+    state: absent
+
+# Install a snap with classic confinement
+- name: Install "foo" with option --classic
+  snap:
+    name: foo
+    classic: True
+'''
+
+RETURN = '''
+msg:
+    description: An informative message on the task's output
+    type: str
+    returned: on error
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def clean_err(err):
+    # Remove whitespaces and newlines that the snap CLI outputs.
+    return err.replace('\n', '').replace('      ', '').replace('"', '\'')
+
+
+def snap_exists(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'info', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc, out, err
+
+
+def is_snap_installed(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'list', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc, out, err
+
+
+def install_snaps(module, snap_names, classic):
+    snaps_to_install = list()
+    for snap_name in snap_names:
+        rc, out, err = is_snap_installed(module, snap_name)
+        if rc != 0:
+            # Snap is not installed
+            snaps_to_install.append(snap_name)
+
+    if not snaps_to_install:
+        snaps_already_installed = ', '.join(snap_names)
+        module.exit_json(msg="Snap(s) already installed: %s" % str(snaps_already_installed), classic=classic, changed=False)
+    else:
+        # Transform the list into a string with whitespace-separated snaps
+        snaps_to_install = ' '.join(snaps_to_install)
+        # Create a string with commas for the output
+        snaps_installed = snaps_to_install.replace(' ', ', ')
+
+    if module.check_mode:
+        module.exit_json(msg="Snap(s) that would have been installed: %s" % str(snaps_installed), classic=classic, changed=True)
+
+    classic = '--classic' if classic else ''
+
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'install', snaps_to_install, classic]
+    cmd = ' '.join(cmd_parts)
+
+    # Actually install the snaps
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    if rc == 0:
+        module.exit_json(msg="Snap(s) installed: %s" % str(snaps_installed), classic=classic, changed=True)
+    else:
+        err = clean_err(err)
+        module.fail_json(msg="Something went wrong.", classic=classic, err=err)
```

Comment:
You could use some matching like this to provide better error message, then "something went horribly wrong"
```python
m = re.match(r'^error: This revision of snap "(?P<package_name>\w+)" was published using classic confinement and thus may perform')
if m is not None:
    err_pkg = m.group('package_name')
    err = "Couldn't install {name} because it uses classic confinement".format(name=err_pkg)
```

stderr details are available to user anyway, when they use `-vvv` in `ansible-playbook` command or through registered result, so you don't have to return that manually. Try it out to see how it works.

Reply:
Works well!

https://github.com/ansible/ansible/pull/44939/commits/21c131d1e95d1a65723de1464f5f67d790b5276d

Sample:

```yml
fatal: [user@host]: FAILED! => {
    "changed": false,
    "classic": false,
    "cmd": "/usr/bin/snap install micro ",
    "invocation": {
        "module_args": {
            "classic": false,
            "name": [
                "micro"
            ],
            "state": "present"
        }
    },
    "msg": "Couldn't install micro because it uses classic confinement",
    "stderr": "error: This revision of snap \"micro\" was published using classic confinement\n       and thus may perform arbitrary system changes outside of the security\n       sandbox that snaps are usually confined to, which may put your system at\n       risk.\n\n       If you understand and want to proceed repeat the command including\n       --classic.\n",
    "stderr_lines": [
        "error: This revision of snap \"micro\" was published using classic confinement",
        "       and thus may perform arbitrary system changes outside of the security",
        "       sandbox that snaps are usually confined to, which may put your system at",
        "       risk.",
        "",
        "       If you understand and want to proceed repeat the command including",
        "       --classic."
    ],
    "stdout": "",
    "stdout_lines": []
}
```